dawn/src/dawn/Optimizer/PassInlining.cpp

```diff
@@ -388,14 +384,14 @@ class DetectInlineCandiates : public iir::ASTVisitorForwarding {
       : strategy_(strategy), instantiation_(instantiation), inlineCandiatesFound_(false) {}
 
   /// @brief Process the given statement
-  void processStatment(const std::unique_ptr<iir::StatementAccessesPair>& stmtAccesesPair) {
+  void processStatment(std::unique_ptr<iir::StatementAccessesPair>& stmtAccesesPair) {
     // Reset the state
     inlineCandiatesFound_ = false;
-    oldStmtAccessesPair_ = stmtAccesesPair->clone();
+    oldStmtAccessesPair_ = &stmtAccesesPair;
     newStmtAccessesPairs_.clear();
 
     // Detect the stencil functions suitable for inlining
-    oldStmtAccessesPair_->getStatement()->accept(*this);
+    (*oldStmtAccessesPair_)->getStatement()->accept(*this);
```

Comment:
See line 406.
The idea here is to steal the `unique_ptr` (pointed by `oldStmtAccessesPair_`). Most of the time you don't want to steal, but sometimes you do.
This happens in 2 phases: in `processStatement` you identify a candidate and if `getNewStatementAccessesPairs` is executed you move the `unique_ptr`.

Reply:
I see the point. I think again, this is very complex code flow, but this is not your fault because it was already before. It looks like a weird pattern, that this class has three functions, because the three functions are always used together and in the same order (`c.process(); if (c.hasResult()) ret = c.getResult();`). For this three functions we have some internal state like `inlineCandiatesFound_`, `oldStmtAccessesPair_` and `newStmtAccessesPairs_`, but there is also some state about several calls to these three functions.

So *if* you want to clean it up a bit, I would try to merge the three functions. But you don't need to do it, because it was already like this before.

Reply:
Somehow I like the idea of separating the pass into an analysis step that determines the operation to do (inline or don't) and an "apply decision" step. So i would still keep the separation, but find a way to improve the information passing from the first step to second ... in this case `std::unique_ptr<iir::StatementAccessesPair>*` could be substituted by some other way of referencing to the statement, such as an index over the children of the current `doMethod` or an iterator?

Reply:
I propose if we want to change the current implementation, we can do it in a new PR because this one is already pretty large. That's then also easier to discuss. Do you think that's a reasonable approach?

Reply:
Yes that's reasonable, I'll open an issue on Github after the PR is merged